dark/consensus.py

```diff
@@ -1,102 +1,88 @@
+    @param minCoverage: An C{int} minimum number of reads that must cover a
+        site for a threshold consensus base to be called. If zero reads
+        cover a site, the C{noCoverage} value is used or if the number is
+        greater than zero but less then C{minCoverage}, the C{lowCoverage}
+        value is used.
+    @param lowCoverage: A C{str} indicating what to do when some reads cover a
+        site, but fewer than C{minCoverage}. Either 'reference' or a single
+        character (e.g., 'N').
+    @param noCoverage: A C{str} indicating what to do when no reads cover a
+        reference base. Either 'reference' or a single character (e.g., 'N').
+    @param ignoreQuality: If C{True}, ignore quality scores.
+    @return: A C{str} consensus sequence.
+    """
+    with samfile(bamFilename) as fp:
+        if strategy == 'majority':
+            return _majorityConsensus(fp, reference, threshold, minCoverage,
+                                      lowCoverage, noCoverage, ignoreQuality)
+        else:
+            raise ValueError(f'Unknown consensus strategy {strategy!r}.')
+
+
+def _majorityConsensus(bam, reference, threshold, minCoverage, lowCoverage,
+                       noCoverage, ignoreQuality):
+    """
+    Compute a majority consensus.
+
+    @param bam: An open BAM file.
+    @param reference: A C{dark.reads.Read} instance giving the reference
         sequence.
-    actualSequenceId: the str id of the actual sequence (if known).
+    @param threshold: A C{float} threshold. If the majority ...
+    @param minCoverage: An C{int} minimum number of reads that must cover a
+        site for a threshold consensus base to be called. If zero reads
+        cover a site, the C{noCoverage} value is used or if the number is
+        greater than zero but less then C{minCoverage}, the C{lowCoverage}
+        value is used.
+    @param lowCoverage: A C{str} indicating what to do when some reads cover a
+        site, but fewer than C{minCoverage}. Either 'reference' or a single
+        character (e.g., 'N').
+    @param noCoverage: A C{str} indicating what to do when no reads cover a
+        reference base. Either 'reference' or a single character (e.g., 'N').
+    @param ignoreQuality: If C{True}, ignore quality scores.
+    @return: A C{str} consensus sequence.
     """
+    result = list(reference.sequence if noCoverage == 'reference' else
+                  noCoverage * len(reference))
+    lowCoverage = (reference.sequence if lowCoverage == 'reference' else
+                   lowCoverage * len(reference))
 
-    print('TODO: This function is not finished yet.')
-    return
+    for column in bam.pileup(reference=reference.id):
+        site = column.reference_pos
+        bases = defaultdict(int)
+        readCount = 0
+        for read in column.pileups:
+            readCount += 1
+            base = read.alignment.query_sequence[read.query_position]
+            bases[base] += (
```

Comment:
i'd find sth like this more readable/explicit: 
`bases[base] += 1 if ignoreQuality else read.alignment.query_qualities[read.query_position]`

Reply:
Yeah, I figured that line was going to get flagged. I originally wrote it pretty much as you suggest, then realized there was a cuter (= more obscure) way to do it...  I'll put it back. Thanks for reading closely :-)